sql: fix missing cleanup

andreimatei opened the pull request:
execStmtInParallel was not calling the cleanup function on early
returns. This was causing a RETURNING NOTHING to potentially stay in
SHOW QUERIES forever if it encountered a planning error.

The planning errors were also not handled correctly wrt communicating to
the client. The caller of execStmtInParallel was not respecting the
contract by not setting the error on the CommandResult. This one did not
actually translate into a user-visible problem, since we had a
higher-level catch-all in the connEx that made sure that if we got an
error event, we also set an error on the CommandResult.

Release note: None

andreimatei committed:
sql: fix missing cleanup

execStmtInParallel was not calling the cleanup function on early
returns. This was causing a RETURNING NOTHING to potentially stay in
SHOW QUERIES forever if it encountered a planning error.

The planning errors were also not handled correctly wrt communicating to
the client. The caller of execStmtInParallel was not respecting the
contract by not setting the error on the CommandResult. This one did not
actually translate into a user-visible problem, since we had a
higher-level catch-all in the connEx that made sure that if we got an
error event, we also set an error on the CommandResult.

Release note: None

andreimatei committed:
sql: unindent a bunch of code

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35341)
<!-- Reviewable:end -->

andreimatei committed:
sql: fix missing cleanup on early return

More missing cleanups: ParallelizeQueue.Add() was missing the query
de-registration in an error case when the query execution callback was
not being called.
This patch also generally makes cleanup more robust around the
ParallelizeQueue.

Release note: None

knz reviewed (approved):
Good catch! LGTM, but I wish this parallel execution stuff was gone already.

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35341#-:-L_4huXpEYfn8p7BSlNL:bdy3pme)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/35341#-:-L_8ZqHX9HebJNgOctGS:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->
